### ISSUE TITLE
Failed imports are forever treated as unknown symbols

### ISSUE THREAD

**Issue by Aran-Fey** [microsoft/pyright#4076]
**Describe the bug**
If an import fails, but the same import is then performed from a backport module, Pyright still treats it as an unknown symbol.

**To Reproduce**
```python
try:
    from typing import Self  # Fails because I don't have 3.11
except ImportError:
    from typing_extensions import Self  # Succeeds

class Demo:
    def method(self) -> Self:
        return 3  # No error displayed in this line!
```

**Expected behavior**
Pyright should assume that both imports are equivalent, and not treat `Self` as an unknown symbol.

**Comment by bschnurr** [comment]
try using 

```
import sys

if sys.version_info < (3, 11):

```

**Comment by Aran-Fey** [comment]
Oh, thanks, that's certainly useful. But unfortunately it doesn't really solve my problem. I'm actually importing `dataclass_transform`, and I have 3 branches in total:

```python
import sys

if sys.version_info >= (3, 11):
    from typing import dataclass_transform
else:
    try:
        from typing_extensions import dataclass_transform
    except ImportError:
        dataclass_transform = lambda: (lambda f: f)

@dataclass_transform()
def my_dataclass(cls):
    return cls

@my_dataclass
class Demo:
    x: int

Demo(3)
```

I don't understand why, but this gives `error: Expected no arguments to "Demo" constructor` as output.

And in any case, it would be nice if we didn't have to hard-code specific version numbers.

**Comment by erictraut** [comment]
The current behavior is "as designed". I don't consider this a bug. In cases where a symbol is imported both within a `try` and an `except` block, pyright assumes that it should use the `try`. Obviously, pyright cannot know whether an exception will be raised at runtime, so it needs to apply a heuristic. This current heuristic works for most cases that we've seen.

What you're doing here is atypical in wrapping an import from `typing` in a `try` block. There's no need to do this. If you need to run on older versions of Python that don't support a typing symbol, simply import it from `typing_extensions`. There's no need to attempt to import it from `typing` conditionally. You're making things more difficult for yourself (and for pyright) than necessary. On newer versions of Python, `typing_extensions` will use the corresponding symbol from `typing`.

```python
from typing_extensions import Self
```

**Comment by Aran-Fey** [comment]
You're saying I should depend on `typing_extensions` even if it's not necessary because the user already has python 3.11? I'm not a huge fan of that, but I guess it gets the job done.
